x/shared/session.go

```diff
@@ -106,3 +107,45 @@ func GetProofWindowCloseHeight(sharedParams *sharedtypes.Params, queryHeight int
 	return GetProofWindowOpenHeight(sharedParams, queryHeight) +
 		int64(sharedParams.GetProofWindowCloseOffsetBlocks())
 }
+
+// GetEarliestClaimCommitHeight returns the earliest block height at which a claim
+// for the session that includes queryHeight can be committed for a given supplier
+// and the passed sharedParams.
+func GetEarliestClaimCommitHeight(
+	sharedParams *sharedtypes.Params,
+	queryHeight int64,
+	claimWindowOpenBlockHash []byte,
+	supplierAddr string,
+) int64 {
+	claimWindowOpenHeight := GetClaimWindowOpenHeight(sharedParams, queryHeight)
+
+	// Generate a deterministic random (non-negative) int64, seeded by the claim
+	// window open block hash and the supplier address.
+	randomNumber := poktrand.SeededInt63(claimWindowOpenBlockHash, []byte(supplierAddr))
+
+	distributionWindowSizeBlocks := sharedParams.ClaimWindowCloseOffsetBlocks - sharedParams.ClaimWindowOpenOffsetBlocks
+	randCreateClaimHeightOffset := randomNumber % int64(distributionWindowSizeBlocks)
+
+	return claimWindowOpenHeight + randCreateClaimHeightOffset
+}
+
+// GetEarliestProofCommitHeight returns the earliest block height at which a proof
+// for the session that includes queryHeight can be committed for a given supplier
+// and the passed sharedParams.
+func GetEarliestProofCommitHeight(
```

Comment:
Have you put thought into making this be part of the keeper and just having it query sharedParams?

I'm thinking of a simpler less "functional" approach

Reply:
TL;DR, there is no appropriate object to make these static functions  members of.

Like the other helpers in this package, this static function is intended to be the source of truth (identity) for "earliest proof commit height". These functions are used both on and off chain, so while we could move them into the shared keeper package, that would result in the relay miner's dependency try including on-chain code. This would likely bloat the relayminer binary size.

For the sake of completeness, I also considered moving these functions to become methods of `sharedtypes.Param` but this does't feel right to me either.